scripts/new_python_based_image.py

```diff
@@ -0,0 +1,555 @@
+        process_pipfiles(path, target_version)
+
+
+def process_pipfiles(path, target_version):
+    """
+    Processes Pipfiles in a given path by running `pipenv lock` on them.
+
+    Args:
+        path (str): The path to search for Pipfiles.
+        target_version (str): The target Python version to use with `pipenv lock`.
+    """
+    for root, _, files in os.walk(path):
+        for file_name in files:
+            if file_name.startswith("Pipfile") and "lock" not in file_name:
+                pipfile_path = os.path.join(root, file_name)
+                run_pipenv_lock(pipfile_path, target_version)
+
+
+def run_pipenv_lock(pipfile_path, target_version):
+    """
+    Runs `pipenv lock` for a specified Pipfile to generate a new lock file.
+
+    Args:
+        pipfile_path (str): The path to the Pipfile.
+        target_version (str): The target Python version to use with `pipenv lock`.
+    """
+    LOGGER.info(f"Running pipenv lock for '{pipfile_path}'")
+    env = os.environ.copy()
+    env["PIPENV_PIPFILE"] = os.path.basename(pipfile_path)
+
+    try:
+        result = subprocess.run(
+            ["pipenv", "lock", "--python", target_version],
+            cwd=os.path.dirname(pipfile_path),
+            check=True,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            env=env
+        )
+        LOGGER.debug(result.stdout.decode())
+    except subprocess.CalledProcessError as e:
+        LOGGER.error(f"Error running pipenv lock for '{pipfile_path}'")
+        LOGGER.debug(e.stderr.decode())
+
+
+def manual_checks():
+    """
+    Provides a list of manual checks to perform after the script execution.
+
+    Returns:
+        list: A list of manual checks to review.
+    """
+    return [
+        "Check the issues thrown during the script execution, if any, and fix them manually.",
+        "Check if the Python version replacements have been performed correctly on the new files.",
+        "Review and make the appropriate changes in the Makefile and CI-related files.",
+        "Push the changes to a new branch on your fork to build the new images with GitHub workflows.",
+        "Test the new images."
+    ]
+
+
+def main():
+    """
+    Main function to execute the script.
+    """
+    args = extract_input_args()
+    LOGGER.info(args)
+
+    configure_logger(args.log_level)
+
+    with_logged_execution(lambda: check_requirements(args),
+                          "Checking requirements")
+
+    paths = with_logged_execution(lambda: find_matching_paths(args.context_dir, args.source, args.match),
+                                  f"Finding matching paths with '{args.match}' and Python {args.source}")
+
+    paths_dict = replace_python_version_on_paths(paths,
+                                                 args.source,
+                                                 args.target)
+
+    if len(paths_dict) == 0:
+        LOGGER.info("No paths found to copy.")
+        return
```

Comment:
Shall we exit gracefully or is this rather a reason to exit with error?

Reply:
I don't see it as an error if you provide inputs that lead to nothing being processed.

Reply:
Hm, probably depends on point of view. In case I provide wrong input and didn't realize, the script will finish with 0. But yeah, I don't insist. Keep it that way then.

Reply:
Although there must be always a manual review after the script execution, I'll return non-zero code for non-happy scenarios.

Reply:
Here's example from Python itself. Should a test run that ended up ignoring every test and not actually running any be considered a success or a failure? :P Here's it in pytest, but I also recall there was something like that in the stdlib unittest https://github.com/pytest-dev/pytest/pull/817